text/projects/Lukasz_Selwa/Galil_Seiferas.tex

```diff
@@ -0,0 +1,308 @@
+Jeśli $|u| \in [L_i, R_i]$ to $u$ musi być prefiksem $v_i^e$, dla $e \geq 2$. Z pierwotności $v_i$ najkrótszy okres $u$ to $|v_i| = L_i / 2$.
+\end{proof}
+
+Załóżmy, że $w$ jest wzorcem, dla którego $v_1$ jest jedynym $k$-HRP z zakresem $[L_1, R_1]$. Wtedy możemy znaleźć wszystkie wystąpienia $w$ w tekście $t$ w czasie liniowym.
+Niech $m = |w|, n = |t|$.
+
+\begin{algorithm}[H]
+\caption{SIMPLE-TEXT-SEARCH}
+\label{simple-text-search}
+\begin{algorithmic} 
+\STATE $pos \gets 0, j \gets 0$
+\WHILE{$pos + m \leq n$}
+    \WHILE{$j < m \land w[j + 1] = t[pos + j + 1] $}
+        \STATE $j \gets j + 1$
+    \ENDWHILE
+    \IF{$j = m$}
+        \RETURN Match at position $pos$.
+    \ENDIF
+    \IF{$j \in [L_1, R_1]$}
+        \STATE $pos \gets pos + L_1 / 2$
+        \STATE $j \gets j - L_1 / 2$
+    \ELSE
+        \STATE $pos \gets pos + \lfloor j / k \rfloor + 1$
+        \STATE $j \gets 0$
+    \ENDIF
+\ENDWHILE
+\end{algorithmic}
+\end{algorithm}
+Poprawność algorytmu wynika z lematu \ref{Lem_2}. Algorytm działa w czasie liniowym ponieważ cały czas rośnie wartość $k \cdot pos + j$. Łatwo też zauważyć, że dodatkowa pamięć wymagana przez algorytm jest stała.
+
+\begin{definition}
+\textbf{\textit{k-perfekcyjna-dekompozycja.}} Dla danego $w$ mówimy, że słowa $u,v$ są $k$-perfekcyjną dekompozycją jeśli $w = uv$, $v$ ma co najwyżej jedno $k$-HRP oraz $|u| < 2p(v)$.
+\end{definition}
+
+Zauważmy, że jeśli dla wzorca $w$ mamy $k$-perfekcyjną dekompozycję to możemy łatwo znaleźć wszystkie wystąpienia wzorca w czasie liniowym i stałej pamięci za pomocą algorytmu \ref{simple-text-search}. W czasie liniowym możemy wyszukać wszystkie wystąpienia $v$ i dla każdego sprawdzić czy jest poprzedzane przez $u$.
+Złożoność zamortyzuje się do czasu liniowego ponieważ $|u| \leq 2|v|$.
+Przedstawimy algorytm znajdujący $k$-perfekcyjną dekompozycję.
+
+\subsection{Znajdowanie dekompozycji}
+
+Ustalmy $k \geq 3$. Przez $\HRP i(x)$ oznaczamy $i$-ty najkrótszy $\HRP$ słowa $x$.
+\begin{lemma}\label{Lem_3}
+Załóżmy, że $x$ ma $\HRP$, niech $z=\HRP(x)$ oraz $x = zx'$.
+\begin{enumerate}
+    \item Jeśli istnieje $\HRP2(x)$ to $|\HRP2(x)| > 2\cdot |\HRP1(x)|$
+    \item Jeśli istnieje $\HRP1(x')$ to $\HRP1(x)$ jest prefiksem $\HRP1(x')$
+\end{enumerate}
+\begin{proof}
+Załóżmy nie wprost, że $|\HRP2(x)| \leq 2\cdot |\HRP1(x)|$. Ponieważ z definicji $\HRP1(x)^2$ jest prefiksem $x$ to $\HRP2(x)$ jest prefiksem $\HRP1(x)^2$. Co przeczy pierwotności $\HRP2(x)$.
+
+Ponieważ $k \geq 3$ to $(\HRP1(x))^2$ jest prefiksem $x'$. Jeśli zachodziłoby $|\HRP1(x')| < |\HRP1(x)|$ to albo $\HRP1(x)$ nie byłoby słowem pierwotnym albo nie byłoby najkrótszym $\HRP$ słowa $x$.
+\end{proof}
+\end{lemma}
+
+\begin{definition}
+\textbf{\textit{Ciąg czynników.}} Dla słowa $x$ ciąg $V(x) = (v_1, v_2, ...)$ definiujemy następująco, $v_1 = \HRP1(x)$. Niech $x'$ będzie słowem takim, że $x = v_1x'$. Wtedy $x_2 = \HRP(x')$ i tak dalej dopóki $\HRP$ istnieją.
```

Comment:
czynników -> podsłów

I know "factors" have two translations, but "czynniki" suggests we deal with numbers, while these are clearly words. Here it's not a problem, but later I got sometimes confused due to that e.g. $v_1 \ldots v_i$ could be interpreted as a concatenation of words or as a product of numbers.